petastorm/pytorch.py

```diff
@@ -13,11 +13,74 @@
 # limitations under the License.
 
 # Must import pyarrow before torch. See: https://github.com/uber/petastorm/blob/master/docs/troubleshoot.rst
-import pyarrow  # noqa: F401 pylint: disable=W0611
+import collections
+import decimal
+import re
 
+import numpy as np
+from six import PY2
 from torch.utils.data.dataloader import default_collate
 
 
+def _sanitize_pytorch_types(row_as_dict):
+    """Promotes values types in a dictionary to the types supported by pytorch. Raises an error if type is clear error
+    if the type can not be promoted.
+
+    The parameter is modified in-place.
+
+    int8, uint16 are promoted to int32; uint32 -> int64;
+    numpy string_, unicode_, object arrays are not supported.
+
+    :param row_as_dict: a dictinoary of key-value pairs. The values types are promoted to pytorch compatible.
+    :return: None
+    """
+    for name, value in row_as_dict.items():
+        # PyTorch supported types are: double, float, float16, int64, int32, and uint8
+        if isinstance(value, np.ndarray):
+            if value.dtype == np.int8:
+                row_as_dict[name] = value.astype(np.int16)
+            elif value.dtype == np.uint16:
+                row_as_dict[name] = value.astype(np.int32)
+            elif value.dtype == np.uint32:
+                row_as_dict[name] = value.astype(np.int64)
+            elif re.search('[SaUO]', value.dtype.str) is not None \
+                    or value.dtype == np.uint64:
+                raise NotImplementedError('%s is an ndarray of type %s. PyTorch does not support tensors of this '
+                                          'type. Convert to one of the supported types using DataLoader.transform.',
+                                          name, value.dtype)
+
+
+if PY2:
+    _string_classes = basestring  # noqa: F821
+else:
+    _string_classes = (str, bytes)
+
+
+def decimal_friendly_collate(batch):
+    """A wrapper on top of ``default_collate`` function that allows decimal.Decimal types to be collated.
+
+    We use ``decimal.Decimal`` types in petastorm dataset to represent timestamps. PyTorch's ``default_collate``
+    implementation does not support collating ``decimal.Decimal`` types. ``decimal_friendly_collate`` collates
+    ``decimal.Decima al`` separately and then combines with the rest of the fields collated by a standard
+    ``default_collate``.
+
+    :param batch: A list of dictionaries to collate
+    :return: A dictionary of lists/pytorch.Tensor types
+    """
+
+    if isinstance(batch[0], decimal.Decimal):
+        return batch
+    elif isinstance(batch[0], collections.Mapping):
+        return {key: decimal_friendly_collate([d[key] for d in batch]) for key in batch[0]}
+    elif isinstance(batch[0], _string_classes):
+        return batch
```

Comment:
Would this case ever occur? (since `_sanitize_pytorch_types` blocks strings)

Reply:
Removed the strings/object blockage from `_sanitize_pytorch_types`. Even before that, string scalars were not triggering an error, only numpy arrays of strings did.